File: ch1/ruby/1-8.rb
#!/usr/bin/env ruby

def cubic_root(x)
    cubic_root_iter(1.0, x, 0)
end

def cubic_root_iter(guess, x, old_guess)
    if good_enough? guess, old_guess then return guess end

    cubic_root_iter(improve(guess,x), x, guess)
end

def good_enough?(guess, old_guess)
    return (guess - old_guess).abs < 0.0001
end

def improve(guess, x)
    return ((x / guess ** 2) + (guess * 2)) / 3
end

result = cubic_root(ARGV[0].to_f)

puts result

File: ch1/js/1-8.js
(function(){
    'use strict';

    function cubicRoot(x) {
        return cubicRootIter(1.0, x, 0);
    }

    function cubicRootIter(guess, x, oldGuess) {
        if (goodEnough(guess, oldGuess) === true) {
            return guess;
        }

        return cubicRootIter(improve(guess, x), x, guess);
    }

    function improve(guess, x) {
        return ((x/square(guess)) + (2 * guess))/3;
    }

    function goodEnough(guess, oldGuess) {
        return Math.abs(guess - oldGuess) < 0.001;
    }

    function square(x) {
        return x * x;
    }

    if (require.main === module) {
        var args, result;
        args = process.argv.slice(2);

        result = cubicRoot(args[0]);

        console.log('result: ' + result);
     }

}());
